### PR TITLE
Supporting partial permissions in entity title service.

### PR DESCRIPTION
**Pull request by dennisoelkers**
## Motivation and Context

This PR is implementing functionality to support returning titles for collections, where a user has access only to a subset of documents.

/nocl No user-facing change.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog2/database/suggestions/MongoEntitySuggestionService.java [hunk 1]
```diff
@@ -56,42 +57,60 @@ public EntitySuggestionResponse suggest(final String collection,
...
                 .find(bsonFilter)
                 .projection(Projections.include(valueColumn))
-                .sort(Sorts.ascending(valueColumn))
-                .limit(perPage)
-                .skip((page - 1) * perPage);
+                .sort(Sorts.ascending(valueColumn));
+
+        final var userCanReadAllEntities = hasAllPermission(subject) || hasReadPermissionForWholeCollection(subject, collection);
+        final var skip = (page - 1) * perPage;
+        final var checkPermission = createPermissionCheck(subject, collection);
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog2/database/suggestions/MongoEntitySuggestionService.java:73]: It is a premature call - this creation is only needed for "I do not have access to all entities" path.

File: graylog2-server/src/main/java/org/graylog2/database/suggestions/MongoEntitySuggestionService.java [hunk 2]
```diff
@@ -56,42 +57,60 @@ public EntitySuggestionResponse suggest(final String collection,
...
+
+    private Stream<Document> paginateWithPermissionCheck(FindIterable<Document> result, int limit, int skip, Predicate<Document> checkPermission) {
+        return MongoUtils.stream(result)
+                .filter(checkPermission)
+                .limit(limit)
+                .skip(skip);
+    }
+
+    private Stream<Document> mongoPaginate(FindIterable<Document> result, int limit, int skip) {
+        return MongoUtils.stream(result.limit(limit).skip(skip));
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog2/database/suggestions/MongoEntitySuggestionService.java:106]: Cannot we use limit and skip directly on the FindIterable, without MongoUtils.stream() call?

Comment (dennisoelkers) [reply graylog2-server/src/main/java/org/graylog2/database/suggestions/MongoEntitySuggestionService.java:106]: That is exactly what we are doing here?

File: graylog2-server/src/main/java/org/graylog2/database/suggestions/MongoEntitySuggestionService.java (hunk 1 above)
Comment (dennisoelkers) [reply graylog2-server/src/main/java/org/graylog2/database/suggestions/MongoEntitySuggestionService.java:73]: Construction of it is very cheap, I am constructing it here so I can reuse it later in the counts query.

**Review by luk-kaminski — APPROVED** [review]
While I do not like the fact that for suggestions, which are supposed to be a quick call, we may end up streaming the whole collection of documents in Mongo (even twice, the second time doing it to get the total count...), I understand it is a known limitation of our whole architecture and I understand that feature-wise it is needed and we have to live with it.

I am very grateful that you have left the Mongo-only path for users that have wider permissions.

File: graylog2-server/src/main/java/org/graylog2/database/suggestions/MongoEntitySuggestionService.java [hunk 3]
```diff
@@ -56,42 +57,60 @@ public EntitySuggestionResponse suggest(final String collection,
...
                                 doc.getString(valueColumn)
                         )
                 )
-                .into(new ArrayList<>());
+                .toList();
 
-        final long total = mongoCollection.countDocuments(bsonFilter);
+        final long total = userCanReadAllEntities
+                ? mongoCollection.countDocuments(bsonFilter)
+                : MongoUtils.stream(mongoCollection.find(bsonFilter)).filter(checkPermission).count();
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog2/database/suggestions/MongoEntitySuggestionService.java:90]: I am not sure if projection won't be a good idea here as well. Entities have many fields, we care only about id in permission check, don't we?

Comment (dennisoelkers) [reply graylog2-server/src/main/java/org/graylog2/database/suggestions/MongoEntitySuggestionService.java:90]: You are right!

**Pull request merged**